docs/CONTRIBUTING.md

```diff
@@ -61,6 +68,32 @@ Do not cast references to raw pointers where it's not needed. Since references c
 
 Generally, a `rcl` type should have a `Drop` impl that calls its `fini()` function. However, there are some cases where this is not possible, e.g. when the type is part of another `rcl` type that owns and finalizes it, or when the `fini()` function takes extra arguments. In these cases, it needs to be decided individually if and where to call `fini()`, and care must be taken to not return early from a function with `?` before the relevant `fini()` functions have been called.
 
+
+## Consistency with major ROS 2 client libraries
+Consistency with the two major ROS 2 client libraries, `rclcpp` and `rclpy`, is a goal of this project – and in particular `rclcpp`.
+
+That means that
+- Roughly the same concepts should exist in all client libraries, e.g. services, parameters, etc.
+- Types, functions, argument names etc. should have the same or similar names as their corresponding items in `rclcpp`/`rclpy` where possible
+- Code should be structured into similar packages by default, e.g. `rosidl_generator_<lang>`, `rcl<lang>` etc.
+- Features should be built on top of `rcl` by default, instead of re-implementing them from scratch
+
+This does _not_ mean that
+- No variability between languages is allowed. What is easy to do, and what is considered idiomatic, is quite different between languages, and thus ROS 2 allows for considerable variability between client languages.
+- Features should always be ported 1:1. Do not rely on the assumption that the best implementation for `rclrs` is whatever `rclcpp` does, but understand the motivation and status of the implementation in the other client libraries. For instance, sometimes a design is implemented incompletely by a client library, the feature is later found to have problems (e.g. the deadlocking of sync client calls in Python), or the client library is constrained by backwards compatibility.
+- Internal consistency does not matter. It does.
```

Comment:
I'd add external consistency as well, API should follow the principle of least surprise, where there are no quirks in how similar functions are called and the types they return.

Reply:
I meant more in the sense of consistence in the `rclrs` API itself, over the years `rclcpp` and `rclpy` have their own quirks and sometimes the API is not entirely coherent. For example, `rclcpp`'s [create_subscription](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/node.hpp#L221-L222) has the QoS parameter before the callback, where [create_service](https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/node.hpp#L316-L317) has them swapped. It's a delicate line, because on the one hand we want to make it easy for people to port their applications and mimic the existing APIs, but if such APIs are not consistent, it may make writing new applications more confusing.

Reply:
In my opinion, if we have to make a decision to mimic `rclcpp`'s API or `rclpy`'s API, we should probably go for `rclcpp` - Rust is a bit closer to C++ than to Python, so (theoretically) it's more likely that a user of `rclcpp` would be using `rclrs`. At least that's my reasoning.

But yeah, ideally the APIs would be consistent across the board.